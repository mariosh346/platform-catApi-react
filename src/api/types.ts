export interface CatImage {
  id: string
  url: string
  breeds?: Breed[]
}

export interface Breed {
  id: string
  name: string
  description: string
}
