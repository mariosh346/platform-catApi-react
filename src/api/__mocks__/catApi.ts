import { vi } from 'vitest'

export const getRandomImages = vi.fn().mockResolvedValue([])
