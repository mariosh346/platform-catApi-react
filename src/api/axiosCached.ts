import axios from 'axios'
import { setupCache } from 'axios-cache-interceptor'

const BASE_URL = 'https://api.thecatapi.com/v1'
const axiosInstance = axios.create({ baseURL: BASE_URL })

export const getAxios = () => setupCache(axiosInstance, { ttl: 60 * 60 * 1000 })