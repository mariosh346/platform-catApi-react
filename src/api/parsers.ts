import { CatImage, Breed } from './types'

export const parseCatImages = (data: unknown): CatImage[] => {
  if (!Array.isArray(data)) {
    return []
  }
  return data.map((item: unknown) => {
    if (typeof item !== 'object' || item === null || !('id' in item) || !('url' in item)) {
      throw new Error("Invalid data: Expected an object with id, name and description for Breed")
    }
    return {
      id: String(item.id),
      url: String(item.url),
      breeds: 'breeds' in item ? parseBreeds(item.breeds) : undefined,
    }
  })
}
export const parseBreed = (item: unknown): Breed => {
  if (
    typeof item !== 'object' ||
    item === null ||
    !('id' in item) ||
    !('name' in item) ||
    !('description' in item)
  ) {
    throw new Error("Invalid data: Expected an object with id, name and description for Breed")
  }

  return {
    id: String(item.id),
    name: String(item.name),
    description: String(item.description),
  }
}



export const parseBreeds = (data: unknown): Breed[] => {
  if (!Array.isArray(data)) {
    return []
  }
  return data.map((item: unknown) => 
    parseBreed(item)
)
}

