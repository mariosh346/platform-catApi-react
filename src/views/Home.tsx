import { JSX, useEffect } from 'react';
import useFetchCatImages from '../hooks/useFetchCatImages';
import ImageGallery from '../components/ImageGallery';
import Button from '../components/atoms/Button';

function Home(): JSX.Element {
  const { images, fetchImages, isLoading, error } = useFetchCatImages();

  useEffect(() => {
    if (isLoading || images.length > 0) return;
    void fetchImages();
    // we only want to fetch new images when the component mounts
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <div>
      <h1>Random Cats</h1>
      {error && <p className="text-red-500 text-center my-4">{error}</p>}
      {images.length === 0 && !isLoading && !error && (
        <p className="text-center my-4">No images found.</p>
      )}
      <ImageGallery images={images} isLoading={isLoading} />
      <div className="flex justify-center my-4">
        <Button
          onClick={() => {
            void fetchImages(false);
          }}
          isLoading={isLoading}
          disabled={isLoading}
        >
          Load More
        </Button>
      </div>
    </div>
  );
}

export default Home
