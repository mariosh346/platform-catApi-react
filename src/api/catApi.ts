import axios from 'axios'
import { parseCatImages, parseBreeds } from './parsers'
import { CatImage, Breed } from './types'

const BASE_URL = 'https://api.thecatapi.com/v1'

export const getRandomImages = async (limit = 10): Promise<CatImage[]> => {
  const res = await axios.get(`${BASE_URL}/images/search?limit=${limit.toString()}`)
  return parseCatImages(res.data)
}

export const getBreeds = async (): Promise<Breed[]> => {
  const res = await axios.get(`${BASE_URL}/breeds`)
  return parseBreeds(res.data)
}

export const getBreedImages = async (breedId: string, limit = 10): Promise<CatImage[]> => {
  const res = await axios.get(`${BASE_URL}/images/search?breed_ids=${breedId}&limit=${limit.toString()}`)
  return parseCatImages(res.data)
}
