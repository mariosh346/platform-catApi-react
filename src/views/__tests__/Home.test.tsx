import { act, render } from '@testing-library/react'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import Home from '../Home'
import * as catApi from '../../api/catApi'


describe('Home', () => {
  const mockGetRandomImages = vi.spyOn(catApi, 'getRandomImages')
  mockGetRandomImages.mockResolvedValue([])
  let rendered: ReturnType<typeof render>;

  beforeEach(() => {
    act(() => {
      rendered = render(
          <Home />
      );
    });
  })

  it('should call getRandomImages once on mount', () => {
    expect(mockGetRandomImages).toHaveBeenCalledTimes(1)
    expect(mockGetRandomImages).toHaveBeenCalledWith(10)
  })
  it('should have the same snapshot as befoere', () => {
    expect(rendered.baseElement).toMatchSnapshot()
  })
})
