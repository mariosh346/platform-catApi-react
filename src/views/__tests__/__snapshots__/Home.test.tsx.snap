// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > should have the same snapshot as befoere 1`] = `undefined`;
